File: gui_view/src/main/scala/com/goexp/galgame/gui/view/game/CVInfoController.java
package com.goexp.galgame.gui.view.game;

import com.goexp.galgame.common.model.CV;
import com.goexp.galgame.gui.task.CVListTask;
import com.goexp.galgame.gui.task.TaskService;
import com.goexp.galgame.gui.util.Tags;
import com.goexp.galgame.gui.util.res.LocalRes;
import com.goexp.galgame.gui.view.DefaultController;
import javafx.collections.ObservableList;
import javafx.concurrent.Service;
import javafx.fxml.FXML;
import javafx.scene.control.Hyperlink;
import javafx.scene.control.TableCell;
import javafx.scene.control.TableColumn;
import javafx.scene.control.TableView;
import javafx.scene.control.cell.PropertyValueFactory;
import javafx.scene.image.ImageView;
import javafx.scene.layout.HBox;

import java.time.LocalDate;
import java.util.List;

public class CVInfoController extends DefaultController {

    @FXML
    private TableView<CV> tableCV;

    @FXML
    private TableColumn<CV, String> colName;
    @FXML
    private TableColumn<CV, Integer> colStar;
    @FXML
    private TableColumn<CV, List<String>> colTag;
    @FXML
    private TableColumn<CV, LocalDate> colStart;
    @FXML
    private TableColumn<CV, LocalDate> colEnd;
    @FXML
    private TableColumn<CV, Integer> colSize;


    private final Service<ObservableList<CV>> loadCVService = new TaskService<>(CVListTask::new);


    protected void initialize() {

        colName.setCellValueFactory(new PropertyValueFactory<>("name"));
        colStar.setCellValueFactory(new PropertyValueFactory<>("star"));
        colStart.setCellValueFactory(new PropertyValueFactory<>("start"));
        colEnd.setCellValueFactory(new PropertyValueFactory<>("end"));
        colSize.setCellValueFactory(new PropertyValueFactory<>("size"));

        colTag.setCellValueFactory(new PropertyValueFactory<>("tag"));

        colStar.setCellFactory(col -> {
            final var image = LocalRes.HEART_16_PNG().get();
            return new TableCell<>() {
                protected void updateItem(Integer star, boolean empty) {
                    super.updateItem(star, empty);
                    this.setGraphic(null);
                    this.setText(null);


                    if (star != null && !empty) {

                        if (star > 0) {
                            var starBox = new HBox();
                            for (var i = 0; i < star; i++) {
                                starBox.getChildren().add(new ImageView(image));
                            }

                            this.setGraphic(starBox);

                        } else {
                            this.setText(star.toString());
                        }
                    }
                }
            };
        });
        colName.setCellFactory(col -> new TableCell<>() {
            @Override
            protected void updateItem(String item, boolean empty) {
                super.updateItem(item, empty);
                this.setGraphic(null);
                this.setText(null);

                if (item != null && !empty) {
                    var link = new Hyperlink(item);
                    link.setOnAction(event -> HomeController.$this().loadCVTab(item, true));
                    this.setGraphic(link);
                }
            }
        });

        colStart.setCellFactory(col -> new TableCell<>() {
            @Override
            protected void updateItem(LocalDate item, boolean empty) {
                super.updateItem(item, empty);

                this.setText(null);
                this.setGraphic(null);

                if (item != null && !empty)
                    this.setText(String.valueOf(item.getYear()));
            }
        });

        colEnd.setCellFactory(col -> new TableCell<>() {
            @Override
            protected void updateItem(LocalDate item, boolean empty) {
                super.updateItem(item, empty);

                this.setText(null);
                this.setGraphic(null);

                if (item != null && !empty)
                    this.setText(String.valueOf(item.getYear()));
            }
        });


        colTag.setCellFactory(col -> new TableCell<>() {
            protected void updateItem(List<String> tag, boolean empty) {
                super.updateItem(tag, empty);
                this.setGraphic(null);
                this.setText(null);

                if (tag != null && !empty) {

                    var hbox = new HBox();
                    hbox.setSpacing(5);
                    hbox.getChildren().setAll(Tags.toNodes(tag));
                    this.setGraphic(hbox);
                }
            }
        });


        loadCVService.valueProperty().addListener((observable, oldValue, newValue) -> {

            if (newValue != null) {
//                var nodes = newValue.stream()
//                        .collect(Collectors.groupingBy(cv -> cv.star))
//                        .entrySet().stream()
//                        .sorted(Comparator.comparing(((Map.Entry entry) -> (Integer) entry.getKey())).reversed())
//                        .map(entry -> {
//                            var flow = new FlowPane();
//                            var panel = new TitledPane(entry.getKey().toString(), flow);
//                            panel.setCollapsible(false);
//                            panel.setPrefWidth(1000);
//
//                            var links = entry.getValue().stream()
//                                    .map(cv -> {
//
//                                        var link = new Hyperlink(cv.name);
////                                        link.setStyle("-fx-font-size:18px");
//                                        link.getStyleClass().add("tag");
//
//                                        return link;
//                                    })
//                                    .collect(Collectors.toUnmodifiableList());
//
//                            flow.getChildren().setAll(links);
//
//
//                            return panel;
//                        })
//                        .collect(Collectors.toUnmodifiableList());
//
//                cvFlow.getChildren().setAll(nodes);


                tableCV.setItems(newValue);
            }

        });


    }


    public void load() {

        loadCVService.restart();

    }


}

File: gui_view/src/main/scala/com/goexp/galgame/gui/view/game/CVInfoController.scala
package com.goexp.galgame.gui.view.game

import java.time.LocalDate

import com.goexp.galgame.common.model.CV
import com.goexp.galgame.gui.task.{CVListTask, TaskService}
import com.goexp.galgame.gui.util.Tags
import com.goexp.galgame.gui.util.res.LocalRes
import com.goexp.galgame.gui.view.DefaultController
import javafx.fxml.FXML
import javafx.scene.control.cell.PropertyValueFactory
import javafx.scene.control.{Hyperlink, TableCell, TableColumn, TableView}
import javafx.scene.image.ImageView
import javafx.scene.layout.HBox

import scala.jdk.CollectionConverters._

class CVInfoController extends DefaultController {
  @FXML private var tableCV: TableView[CV] = _
  @FXML private var colName: TableColumn[CV, String] = _
  @FXML private var colStar: TableColumn[CV, Int] = _
  @FXML private var colTag: TableColumn[CV, List[String]] = _
  @FXML private var colStart: TableColumn[CV, LocalDate] = _
  @FXML private var colEnd: TableColumn[CV, LocalDate] = _
  @FXML private var colSize: TableColumn[CV, Int] = _

  final private val loadCVService = TaskService(new CVListTask())

  override protected def initialize() = {
    colName.setCellValueFactory(new PropertyValueFactory[CV, String]("name"))
    colStar.setCellValueFactory(new PropertyValueFactory[CV, Int]("star"))
    colStart.setCellValueFactory(new PropertyValueFactory[CV, LocalDate]("start"))
    colEnd.setCellValueFactory(new PropertyValueFactory[CV, LocalDate]("end"))
    colSize.setCellValueFactory(new PropertyValueFactory[CV, Int]("size"))
    colTag.setCellValueFactory(new PropertyValueFactory[CV, List[String]]("tag"))

    colStar.setCellFactory(_ => {
      val image = LocalRes.HEART_16_PNG

      new TableCell[CV, Int]() {
        override protected def updateItem(star: Int, empty: Boolean) = {
          super.updateItem(star, empty)
          this.setGraphic(null)
          this.setText(null)
          if (!empty)
            if (star > 0) {
              val stars = Range(0, star).to(LazyList).map { _ => new ImageView(image) }.toArray

              val starBox = new HBox(stars: _*)

              this.setGraphic(starBox)
            }
            else this.setText(star.toString)
        }
      }
    })
    colName.setCellFactory(_ => new TableCell[CV, String]() {
      override protected def updateItem(item: String, empty: Boolean) = {
        super.updateItem(item, empty)
        this.setGraphic(null)
        this.setText(null)
        if (item != null && !empty) {
          val link = new Hyperlink(item)
          link.setOnAction(_ => HomeController.$this.loadCVTab(item, true))
          this.setGraphic(link)
        }
      }
    })
    colStart.setCellFactory(_ => new TableCell[CV, LocalDate]() {
      override protected def updateItem(item: LocalDate, empty: Boolean) = {
        super.updateItem(item, empty)
        this.setText(null)
        this.setGraphic(null)
        if (item != null && !empty) this.setText(String.valueOf(item.getYear))
      }
    })
    colEnd.setCellFactory(_ => new TableCell[CV, LocalDate]() {
      override protected def updateItem(item: LocalDate, empty: Boolean) = {
        super.updateItem(item, empty)
        this.setText(null)
        this.setGraphic(null)
        if (item != null && !empty) this.setText(String.valueOf(item.getYear))
      }
    })
    colTag.setCellFactory(_ => new TableCell[CV, List[String]]() {
      override protected def updateItem(tag: List[String], empty: Boolean) = {
        super.updateItem(tag, empty)
        this.setGraphic(null)
        this.setText(null)
        if (tag != null && !empty) {
          val hbox = new HBox
          hbox.setSpacing(5)
          hbox.getChildren.setAll(Tags.toNodes(tag.asJava))
          this.setGraphic(hbox)
        }
      }
    })

    tableCV.itemsProperty().bind(loadCVService.valueProperty())

  }

  def load() = loadCVService.restart()
}